File: backend/app/schemas/assistant.py
"""AI 求职助手请求、会话和消息结构。"""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

from .material import MAX_MATERIAL_TITLE_CHARS

MAX_ASSISTANT_MESSAGE_CHARS = 20_000
MAX_ATTACHMENT_DATA_CHARS = 7_100_000


class AssistantAttachmentInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=255)
    mime_type: str = Field(default="", max_length=100)
    data: str = Field(min_length=1, max_length=MAX_ATTACHMENT_DATA_CHARS)

    @field_validator("name")
    @classmethod
    def normalize_name(cls, value: str) -> str:
        value = value.strip()
        if not value or any(ord(char) < 32 for char in value):
            raise ValueError("附件名称无效")
        return value


# 思考强度："" 表示不发送该参数（沿用服务商默认），其余透传给支持推理的模型。
REASONING_EFFORTS = ("", "none", "low", "medium", "high")
ReasoningEffort = Literal["", "none", "low", "medium", "high"]


class AssistantMessageCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    content: str = Field(default="", max_length=MAX_ASSISTANT_MESSAGE_CHARS)
    job_id: int | None = Field(default=None, ge=1)
    resume_id: int | None = Field(default=None, ge=1)
    # 保留字段是为了兼容旧版客户端，但现在助手默认读取本地完整个人资料，前端不再提供
    # 关闭开关。无论旧客户端传 false 还是新客户端省略，都在校验后归一为 True。
    include_profile: bool = True
    web_search: bool = False
    # 有思考模式的大模型可以在这里调整推理强度；不支持该参数的服务商会被忽略。
    reasoning_effort: ReasoningEffort = ""
    # 「引用某条消息追问」：指向同一会话里的某条消息，模型会在引用上下文中作答。
    quoted_message_id: int | None = Field(default=None, ge=1)
    attachments: list[AssistantAttachmentInput] = Field(default_factory=list, max_length=4)

    @model_validator(mode="after")
    def require_content_or_attachment(self):
        self.content = self.content.strip()
        self.include_profile = True
        if not self.content and not self.attachments:
            raise ValueError("请输入问题或添加附件")
        return self


class ChatConversationCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str = Field(default="", max_length=120)
    # 为真时自动附上一条引导消息（介绍助手功能与用法），用于首次进入创建默认会话。
    welcome: bool = False


class ChatConversationUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, min_length=1, max_length=120)
    pinned: bool | None = None
    favorite: bool | None = None
    archived: bool | None = None
    # 分组名（"移动到项目"）；空串表示移出分组。
    group_name: str | None = Field(default=None, max_length=64)

    @model_validator(mode="after")
    def require_update(self):
        if all(
            value is None
            for value in (self.title, self.pinned, self.favorite, self.archived, self.group_name)
        ):
            raise ValueError("至少提供一个要修改的会话字段")
        return self

    @field_validator("title")
    @classmethod
    def title_must_not_be_blank(cls, value: str | None) -> str | None:
        if value is None:
            return None
        value = " ".join(value.split())
        if not value:
            raise ValueError("会话标题不能为空")
        return value

    @field_validator("group_name")
    @classmethod
    def group_name_must_be_clean(cls, value: str | None) -> str | None:
        if value is None:
            return None
        return " ".join(value.split())[:64]


class ChatConversationForkRequest(BaseModel):
    """「在新对话中继续」：把原会话最近若干条消息复制成一段新会话。"""

    model_config = ConfigDict(extra="forbid")

    title: str = Field(default="", max_length=120)
    message_limit: int = Field(default=10, ge=1, le=40)


class ChatMessageDeleteRequest(BaseModel):
    """批量删除消息（一次最多 200 条）。"""

    model_config = ConfigDict(extra="forbid")

    message_ids: list[int] = Field(min_length=1, max_length=200)

    @field_validator("message_ids")
    @classmethod
    def ids_must_be_unique(cls, value: list[int]) -> list[int]:
        return list(dict.fromkeys(value))


class ChatMessageDeleteResult(BaseModel):
    deleted: int


class ConversationToMaterialRequest(BaseModel):
    """把一段对话存进资料箱（内容用导出的 Markdown，助手之后能直接读它）。"""

    model_config = ConfigDict(extra="forbid")

    title: str = Field(default="", max_length=MAX_MATERIAL_TITLE_CHARS)
    category: str = Field(default="", max_length=64)
    note: str = Field(default="", max_length=1000)


class ChatAttachmentOut(BaseModel):
    name: str
    mime_type: str
    # document：PDF/DOCX 等文档，文字已在本地提取进 ``text``，原始文件不外发。
    kind: Literal["text", "image", "document"]
    size_bytes: int
    text: str = ""
    data_url: str = ""
    # 提取过程中的说明（例如内容过长只取了前一部分）；目前只有文档会产生。
    notes: list[str] = Field(default_factory=list)


class ChatMessageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    conversation_id: int
    role: Literal["user", "assistant"]
    content: str
    # 引用的消息 id（被引用消息删除后为 None，但 context.quoted 里仍有快照）。
    quoted_message_id: int | None = None
    attachments: list[ChatAttachmentOut] = Field(default_factory=list)
    context: dict[str, Any] = Field(default_factory=dict)
    status: Literal["pending", "complete", "error", "cancelled"]
    error: str = ""
    model: str = ""
    created_at: datetime


class ChatConversationBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    pinned: bool
    favorite: bool
    archived: bool = False
    group_name: str = ""
    # 列表里展示条数与最后活动时间，方便区分同名会话。
    message_count: int = 0
    created_at: datetime
    updated_at: datetime


class ChatConversationDetail(ChatConversationBrief):
    messages: list[ChatMessageOut] = Field(default_factory=list)

File: frontend/src/types/assistant.ts
/** AI 求职助手会话、消息和流式事件类型。 */

export interface AssistantConversationBrief {
  id: number;
  title: string;
  pinned: boolean;
  favorite: boolean;
  /** 已归档的会话默认收进「已归档」筛选，不参与置顶排序。 */
  archived: boolean;
  /** 分组名（"移动到项目"）；空串表示未分组。 */
  group_name: string;
  message_count: number;
  created_at: string;
  updated_at: string;
}

/** 会话列表筛选：全部 / 收藏 / 已归档。 */
export type ConversationFilter = "all" | "favorite" | "archived";

/** 思考强度：空串表示不发送该参数，沿用服务商默认。 */
export type ReasoningEffort = "" | "none" | "low" | "medium" | "high";

export const REASONING_EFFORT_OPTIONS: { value: ReasoningEffort; label: string }[] = [
  { value: "", label: "默认" },
  { value: "none", label: "关闭" },
  { value: "low", label: "低" },
  { value: "medium", label: "中" },
  { value: "high", label: "高" },
];

export interface AssistantConversationForkPayload {
  title?: string;
  message_limit?: number;
}

export interface AssistantAttachment {
  name: string;
  mime_type: string;
  /** document：pdf/docx，正文已由后端在本机提取进 `text`。 */
  kind: "text" | "image" | "document";
  size_bytes: number;
  text: string;
  data_url: string;
  /** 提取过程中的说明（例如内容过长只取了前一部分）。 */
  notes: string[];
}

export interface AssistantSource {
  title: string;
  url: string;
  snippet: string;
}

/**
 * 一次回答里 [来源N] 的「编号 → url」映射条目。
 *
 * 后端在来源**首次收集时**分配全局单调递增编号并持久化到助手消息的
 * `context.source_map`；前端渲染正文里的 [来源N] 时按这个映射解析链接，
 * 而不是拿"去重后的参考来源"数组下标去对——那会跳错来源。
 */
export interface AssistantSourceNumber {
  number: number;
  url: string;
}

/** 助手调用过一次工具的记录；写在助手消息的 context 里，供历史回看。 */
export interface AssistantToolCall {
  name: string;
  arguments: Record<string, unknown>;
  summary: string;
  link: string;
  ok: boolean;
  error: string;
  /**
   * 这次调用是否真的改动了用户的数据（后端 `ToolResult.changed`）。
   *
   * 可选是因为它随功能一起加的：更早存下的历史消息里没有这个字段，缺失按"没改动"处理，
   * 不能据此**主动**声称改动过——这类信息只有后端说了才算数。折叠标题里的「改动了 N 项」
   * 就靠它，所以前端不再自己按工具名猜哪些是写操作（猜一份就会与后端漂移）。
   */
  changed?: boolean;
}

/** 被引用的那条消息的快照：原消息删掉之后这里仍然可读。 */
export interface AssistantQuotedMessage {
  id: number;
  role: "user" | "assistant";
  excerpt: string;
}

export interface AssistantMessage {
  id: number;
  conversation_id: number;
  role: "user" | "assistant";
  content: string;
  /** 引用追问：这条消息引用的是哪一条。 */
  quoted_message_id: number | null;
  attachments: AssistantAttachment[];
  context: {
    job_id?: number | null;
    resume_id?: number | null;
    include_profile?: boolean;
    web_search?: boolean;
    sources?: AssistantSource[];
    /** [来源N] 的「编号 → url」映射（可选：更早存下的历史消息里没有它）。 */
    source_map?: AssistantSourceNumber[];
    tool_calls?: AssistantToolCall[];
    /**
     * 模型的思考过程（开启思考强度时才有），供用户点开查看。
     *
     * 后端对它有存储上限，超限会截断并把 `reasoning_truncated` 置真——界面据此在
     * 结尾补一句"已截断"，不假装这就是全部。可选：更早存下的历史消息里没有它。
     */
    reasoning?: string;
    /** 上面的思考过程是否因超过存储上限被截断。 */
    reasoning_truncated?: boolean;
    quoted?: AssistantQuotedMessage;
  };
  status: "pending" | "complete" | "error" | "cancelled";
  error: string;
  model: string;
  created_at: string;
}

export interface AssistantConversationDetail extends AssistantConversationBrief {
  messages: AssistantMessage[];
}

export interface AssistantAttachmentInput {
  name: string;
  mime_type: string;
  data: string;
}

export type AssistantStreamEvent =
  | {
      type: "start";
      user_message_id: number;
      assistant_message_id: number;
      conversation_title: string;
    }
  | { type: "progress"; message: string }
  | {
      type: "sources";
      sources: AssistantSource[];
      source_map?: AssistantSourceNumber[];
      error: string;
    }
  | ({
      type: "tool";
    } & AssistantToolCall)
  | { type: "delta"; text: string }
  | { type: "reasoning"; text: string }
  | { type: "done"; message: AssistantMessage }
  | { type: "error"; message: string };
